pkg/docker/docker.go

```diff
@@ -768,19 +892,21 @@ func (d *stiDocker) RunContainer(opts RunContainerOptions) error {
 		os.Exit(2)
 	}
 	return interrupt.New(dumpStack, removeContainer).Run(func() error {
-		// Attach to the container.
+		var attachError error
+		// Attach to the container on go thread (different than with fsouza, since it provided a non-blocking attach which we don't seem to have with k8s/engine-api)
+		// still preserve the flow of attaching before starting to handle various timing issues encountered in the past, as well as allow for --run option
 		glog.V(2).Infof("Attaching to container %q ...", containerName)
-		attachOpts := opts.asDockerAttachToContainerOptions()
-		attachOpts.Container = container.ID
-		if _, err = d.client.AttachToContainerNonBlocking(attachOpts); err != nil {
-			glog.V(0).Infof("error: Unable to attach to container %q with options %+v: %v", containerName, attachOpts, err)
-			return err
-		}
+		go func() {
+			attachOpts := opts.asDockerAttachToContainerOptions()
+			if attachError = d.kubeDockerClient.AttachToContainer(container.ID, attachOpts, opts.asDockerAttachToStreamOptions()); attachError != nil {
+				glog.V(0).Infof("error: Unable to attach to container %q with options %+v: %v", containerName, attachOpts, attachError)
+			}
+		}()
```

Comment:
Thiught of that initially too but the go routine that does the attach will
block and not update the channel the main thread is waiting on so we end up
in the same boat.

So I can't help but circle back to the hacky poll the err for small but
sufficient amonut to insure the post for the attach did not fail
"immediately"

On Thursday, July 28, 2016, Ben Parees <notifications@github.com> wrote:

> In pkg/docker/docker.go
> <https://github.com/openshift/source-to-image/pull/554#discussion_r72660230>
> :
>
> >  		glog.V(2).Infof("Attaching to container %q ...", containerName)
> > -		attachOpts := opts.asDockerAttachToContainerOptions()
> > -		attachOpts.Container = container.ID
> > -		if _, err = d.client.AttachToContainerNonBlocking(attachOpts); err != nil {
> > -			glog.V(0).Infof("error: Unable to attach to container %q with options %+v: %v", containerName, attachOpts, err)
> > -			return err
> > -		}
> > +		go func() {
> > +			attachOpts := opts.asDockerAttachToContainerOptions()
> > +			if attachError = d.kubeDockerClient.AttachToContainer(container.ID, attachOpts, opts.asDockerAttachToStreamOptions()); attachError != nil {
> > +				glog.V(0).Infof("error: Unable to attach to container %q with options %+v: %v", containerName, attachOpts, attachError)
> > +			}
> > +		}()
>
> channels are the standard way to communicate state between go routines.
> Setup a channel and have the go routine send either the error, or nil,
> through the channel when it's done.
>
> and then in the main code, do a read from the channel to block until you
> have a result from the go routine.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/openshift/source-to-image/pull/554/files/3de0d0b3b36d99748e143c5118a2921e728b766a..41d1fc0a3fccc90faeb93db9e60c4f701e1dbd4b#r72660230>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADbadDO_7B2H6-X2nD482UHaQGZIpJ59ks5qaN-7gaJpZM4JUVnn>
> .
>